
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Simple descriptive statistics &#8212; Visual guide to concepts in population genetics.</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="The complexities of real data" href="real_data_complications.html" />
    <link rel="prev" title="Population genetic data" href="variation_data_intro.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Visual guide to concepts in population genetics.</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../intro.html">
   Welcome to your Jupyter Book
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Classic models of selection and drift at a single locus
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../classic_selection_drift/haploids.html">
   Drift and selection in haploids
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../classic_selection_drift/diploids.html">
   Drift and selection in diploids
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../classic_selection_drift/diploid_fixation.html">
   Fixation in a diploid population
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  The Kingman coalescent
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/tree_structure.html">
   The structure of a tree
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/model_for_data.html">
   A model for data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/trees.html">
   Trees
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/trees_with_recombination.html">
   Trees with recombination
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/mutation_frequencies.html">
   Mutation frequencies
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../kingman_coalsecent/expectaions.html">
   Theoretical expectations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Linked selection
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../classic_selection_drift/sweep_from_new_mutation.html">
   Selective “sweeps” (from new mutations)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Describing genetic variation at the sequence level
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="variation_data_intro.html">
   Population genetic data
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Simple descriptive statistics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="real_data_complications.html">
   The complexities of real data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bioinformatics.html">
   Bioinformatic issues
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Appendix
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../bibliography.html">
   Literature cited
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="../../_sources/markdown/describing_variation_data/descriptive_statistics.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/markdown/describing_variation_data/descriptive_statistics.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/ThorntonLab/popgen-course-material"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/ThorntonLab/popgen-course-material/issues/new?title=Issue%20on%20page%20%2Fmarkdown/describing_variation_data/descriptive_statistics.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/ThorntonLab/popgen-course-material/main?urlpath=tree/book/markdown/describing_variation_data/descriptive_statistics.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Simple descriptive statistics
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-variation-table">
     The variation table.
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#recoding-the-data">
     Recoding the data
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-number-of-variable-sites">
     The number of variable sites
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-number-of-pairwise-differences">
     The number of pairwise differences
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#heterozygosity">
     Heterozygosity
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#homozygosity">
     Homozygosity
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#normalizing-by-the-length-of-a-genomic-region">
     Normalizing by the length of a genomic region
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ancestral-derived-and-minor-major-alleles">
     Ancestral/derived and minor/major alleles
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-site-or-allele-frequency-spectrum">
   The site (or allele) frequency spectrum
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#with-respect-to-the-derived-allele">
     With respect to the derived allele
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#with-respect-to-the-minor-allele">
     With respect to the minor allele
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#correlations-amongst-variable-sites-linkage-disequilibrium">
   Correlations amongst variable sites (“linkage disequilibrium”)
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-first-of-perhaps-many-d-statistic-s">
     The (first of perhaps many)
     <span class="math notranslate nohighlight">
      \(D\)
     </span>
     statistic(s).
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pearson-s-squared-correlation-coefficient">
     Pearson’s squared correlation coefficient
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="cell tag_hide-input docutils container" id="descriptive-statistics">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># import modules</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span>
<span class="kn">from</span> <span class="nn">myst_nb</span> <span class="kn">import</span> <span class="n">glue</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate example data for this chapter</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span>
                          <span class="n">random_seed</span><span class="o">=</span><span class="mi">54321</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mf">10000.0</span><span class="p">,</span>
                          <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.25</span><span class="o">/</span><span class="mf">1e4</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">12345</span><span class="p">,</span>
                           <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">JC69</span><span class="p">(),</span> <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_mutations</span> <span class="o">&gt;</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="simple-descriptive-statistics">
<h1>Simple descriptive statistics<a class="headerlink" href="#simple-descriptive-statistics" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-variation-table">
<h2>The variation table.<a class="headerlink" href="#the-variation-table" title="Permalink to this headline">¶</a></h2>
<p>The most common representation of population genetic data is the <em>variation table</em>.
This table is simply a “spreadsheet” showing the DNA states at positions that are variable amongst our samples.</p>
<p>Below, we see an example of a variation table.
The table has the following features:</p>
<ol class="simple">
<li><p>It is written in a matrix format.</p></li>
<li><p>Rows correspond to individual genomes.</p></li>
<li><p>Columns correspond to genomic positions that vary.</p></li>
<li><p>The elements of the matrix are the DNA state at a certain position in a certain genome.</p></li>
</ol>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">alleles</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="n">gmatrix</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">genotype_matrix</span><span class="p">(</span><span class="n">alleles</span><span class="o">=</span><span class="n">alleles</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">haplotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">()),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unicode_</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">()):</span>
    <span class="n">haplotypes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">alleles</span><span class="p">[</span><span class="n">a</span><span class="p">]</span>
<span class="n">haplotypes</span> <span class="o">=</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[&#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;T&#39;]
 [&#39;C&#39; &#39;G&#39; &#39;C&#39; &#39;G&#39;]]
</pre></div>
</div>
</div>
</div>
<p>The first row corresponds to the DNA states at all variable sites for the first genome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[</span><span class="mi">0</span><span class="p">,:])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;G&#39; &#39;T&#39; &#39;T&#39; &#39;T&#39;]
</pre></div>
</div>
</div>
</div>
<p>By comparing the rows to one another, we can start to see that some genomes are identical and other differ at one or more positions.
There are ten rows, which is the number of genomes in the sample.</p>
<p>The columns show use the states corresponding to each <em>site</em> in each genome.
For example, the first column is:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)))</span>
<span class="n">glue</span><span class="p">(</span><span class="s2">&quot;first_genome_first_site&quot;</span><span class="p">,</span> <span class="n">haplotypes</span><span class="p">[:,</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">display</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">glue</span><span class="p">(</span><span class="s2">&quot;second_genome_first_site&quot;</span><span class="p">,</span> <span class="n">haplotypes</span><span class="p">[:,</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">display</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[&#39;G&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]
 [&#39;G&#39;]
 [&#39;C&#39;]
 [&#39;C&#39;]]
</pre></div>
</div>
</div>
</div>
<p>Starting from the top, the first entry, <span class="pasted-text">G</span>, is the DNA state at the first variable site in the first genome.
The second entry, <span class="pasted-text">C</span>, is the DNA state at the first variable site in the second genome.
Etc..</p>
<p>Our table is missing some important things:</p>
<ol class="simple">
<li><p>The positions are omitted.
How the positions are stored is a bioinformatic issue and will have to consider both the chromosome and the position along the chromosome.</p></li>
<li><p>Sample <em>labels</em> are also missing.
We are being intentionally vague/general here.
For example, the table could consist of ten haploid genomes.
Or, it could consist of five diploid genomes with each pair of rows representing the maternal/paternal genomes of an individual.
In practice, we need row labels in order to make full use of the data.</p></li>
</ol>
<p>A fuller representation of our data may thus look like the following, with columns labelled by position and rows labelled by sample identifier:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">sites</span><span class="p">()):</span>
    <span class="n">data</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">position</span><span class="p">)]</span> <span class="o">=</span> <span class="n">haplotypes</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sample &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">))]</span>

<span class="n">display</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>83</th>
      <th>1064</th>
      <th>2987</th>
      <th>6531</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample 0</th>
      <td>G</td>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 1</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 2</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 3</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 4</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 5</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 6</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 7</th>
      <td>G</td>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 8</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>sample 9</th>
      <td>C</td>
      <td>G</td>
      <td>C</td>
      <td>G</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="recoding-the-data">
<h2>Recoding the data<a class="headerlink" href="#recoding-the-data" title="Permalink to this headline">¶</a></h2>
<p>It is often inconvenient to work with the data in the form shown above.
For reasons having to do with efficient computation, it is often useful to recode the data.
For example, we may choose to replace <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code> with <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code> and <code class="docutils literal notranslate"><span class="pre">3</span></code>, respectively:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">gmatrix</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[2 3 3 3]
 [1 2 1 3]
 [1 2 1 3]
 [1 2 1 3]
 [1 2 1 3]
 [1 2 1 3]
 [1 2 1 3]
 [2 3 3 3]
 [1 2 1 3]
 [1 2 1 2]]
</pre></div>
</div>
</div>
</div>
<p>There is no loss of generality here.
Rather, these recodings are practical matters that you are likely to run across when analyzing data.
Various programs will expect the data to be encoded in a certain way.
Fortunately, some standards are emerging.
See <a class="reference internal" href="bioinformatics.html#variation-data-bioinformatics"><span class="std std-ref">here</span></a>.</p>
</div>
<div class="section" id="the-number-of-variable-sites">
<h2>The number of variable sites<a class="headerlink" href="#the-number-of-variable-sites" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">glue</span><span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">display</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>The simplest summary of the data is a tally of the number of variable positions.
For our example data set, this number is <span class="pasted-text">10</span>.</p>
<p>This statistic can be referred to by many synonyms, including:</p>
<ul class="simple">
<li><p>The number of polymorphic sites</p></li>
<li><p>The number of segregating sites, where “segregation” is a term from Mendelian genetics.
This term is most famously associated with the paper by <span id="id1">Watterson [<a class="reference internal" href="../bibliography.html#id6">1975</a>]</span>.</p></li>
</ul>
</div>
<div class="section" id="the-number-of-pairwise-differences">
<h2>The number of pairwise differences<a class="headerlink" href="#the-number-of-pairwise-differences" title="Permalink to this headline">¶</a></h2>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ndiffs</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gmatrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">gmatrix</span><span class="p">[</span><span class="n">i</span><span class="p">,:],</span> <span class="n">gmatrix</span><span class="p">[</span><span class="n">j</span><span class="p">,:]):</span>
            <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="n">l</span><span class="p">:</span>
                <span class="n">ndiffs</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">ncomps</span> <span class="o">=</span> <span class="n">gmatrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>
<span class="n">ndiffs</span> <span class="o">=</span> <span class="n">ndiffs</span> <span class="o">/</span> <span class="n">ncomps</span>
<span class="n">tsdiv</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">([[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()]],</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">assert</span> <span class="n">math</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">ndiffs</span><span class="p">,</span> <span class="n">tsdiv</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ndiffs</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">tsdiv</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="n">glue</span><span class="p">(</span><span class="s2">&quot;mean_diffs&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">ndiffs</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">display</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>We are often concerned with how different our sample genomes are.
A natural summary of the data then is the mean number of differences between all pairs of genomes.
(Here, a pair never includes the same genome compared to itself).
To calculate this value, one compares each genome to all others in the sample.
For each pair, count the number of differences and divide by the number of pairs to get the mean.
For the example data, this value is <span class="pasted-text">1.2667</span>.</p>
<p>The explicit formula for this statistic is <span class="math notranslate nohighlight">\(\pi = \frac{\sum_{i=1}^{n-1}\sum_{j=i+1}^{n}D_{ij}}{C}\)</span>, where <span class="math notranslate nohighlight">\(D_{ij}\)</span> is the number of differences between genomes <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span> and <span class="math notranslate nohighlight">\(C = {n \choose 2} = \frac{n(n-1)}{2}\)</span> for <span class="math notranslate nohighlight">\(n\)</span> genomes.  The term <span class="math notranslate nohighlight">\(C\)</span> is the number of pairs.</p>
<p>TODO Citations: Nei and Tajima.</p>
</div>
<div class="section" id="heterozygosity">
<h2>Heterozygosity<a class="headerlink" href="#heterozygosity" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">counts_site_0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">counts_site_0</span>

<span class="n">het_site_0</span> <span class="o">=</span> <span class="mf">0.0</span>

<span class="n">nsam</span> <span class="o">=</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts_site_0</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
    <span class="n">het_site_0</span> <span class="o">+=</span> <span class="n">c</span> <span class="o">*</span> <span class="p">(</span><span class="n">nsam</span> <span class="o">-</span> <span class="n">c</span><span class="p">)</span>
<span class="n">het_site_0</span> <span class="o">/=</span> <span class="p">(</span><span class="n">nsam</span> <span class="o">*</span> <span class="p">(</span><span class="n">nsam</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="n">counts_site_0</span><span class="p">,</span> <span class="n">het_site_0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([&#39;C&#39;, &#39;G&#39;], dtype=&#39;&lt;U1&#39;), array([8, 2])) 0.35555555555555557
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="homozygosity">
<h2>Homozygosity<a class="headerlink" href="#homozygosity" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hom_site_0</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts_site_0</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
    <span class="n">hom_site_0</span> <span class="o">+=</span> <span class="p">(</span><span class="n">c</span><span class="o">/</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">*</span><span class="p">((</span><span class="n">c</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>

<span class="k">assert</span> <span class="n">math</span><span class="o">.</span><span class="n">isclose</span><span class="p">(</span><span class="n">het_site_0</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">hom_site_0</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">hom_site_0</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">hom_site_0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.6444444444444445 0.3555555555555555
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="normalizing-by-the-length-of-a-genomic-region">
<h2>Normalizing by the length of a genomic region<a class="headerlink" href="#normalizing-by-the-length-of-a-genomic-region" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="ancestral-derived-and-minor-major-alleles">
<h2>Ancestral/derived and minor/major alleles<a class="headerlink" href="#ancestral-derived-and-minor-major-alleles" title="Permalink to this headline">¶</a></h2>
<p>The ancestral and derived states in our example data are known:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site index </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> has ancestral state </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">ancestral_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site index 0 has ancestral state G
Site index 1 has ancestral state T
Site index 2 has ancestral state T
Site index 3 has ancestral state T
</pre></div>
</div>
</div>
</div>
<p>The derived states are:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> at site index </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">site</span><span class="si">}</span><span class="s2"> has derived state </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">derived_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Mutation 0 at site index 0 has derived state C
Mutation 1 at site index 1 has derived state G
Mutation 2 at site index 2 has derived state C
Mutation 3 at site index 3 has derived state G
</pre></div>
</div>
</div>
</div>
<p>Using the convention of representing the ancestral state as <code class="docutils literal notranslate"><span class="pre">0</span></code> and all derived states as integer values greater than <code class="docutils literal notranslate"><span class="pre">0</span></code>, we can rewrite our data matrix as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">derived_state_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="n">unfolded_fs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">derived_state_map</span><span class="p">:</span>
        <span class="n">derived_state_map</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">derived_state</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">derived_state_map</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">derived_state</span><span class="p">)</span>

<span class="n">recoded_matrix_ancestral_derived</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">haplotypes</span><span class="p">[:,</span><span class="n">site</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">derived_state_map</span><span class="p">[</span><span class="n">site</span><span class="p">]:</span>
            <span class="n">recoded_matrix_ancestral_derived</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">derived_state_map</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">state</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">recoded_matrix_ancestral_derived</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>


<span class="c1"># This is some ugly business.</span>
<span class="c1"># We are recoding the input *columns* rather than the rows.</span>
<span class="c1"># To get the right matrix out, we need to reverse the shape and</span>
<span class="c1"># then transpose.  This is *not* an example of doing things efficiently.</span>
<span class="n">recoded_matrix_ancestral_derived</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">recoded_matrix_ancestral_derived</span><span class="p">,</span>
                                            <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>

<span class="nb">print</span><span class="p">(</span><span class="n">recoded_matrix_ancestral_derived</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[0 0 0 0]
 [1 1 1 0]
 [1 1 1 0]
 [1 1 1 0]
 [1 1 1 0]
 [1 1 1 0]
 [1 1 1 0]
 [0 0 0 0]
 [1 1 1 0]
 [1 1 1 1]]
</pre></div>
</div>
</div>
</div>
<p>We can similarly recode the data so that <code class="docutils literal notranslate"><span class="pre">0</span></code> is the <em>minor</em> allele and all non-zero integers are <em>major</em> alleles.
For each site in our data, we have:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">minor_allele_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span> <span class="n">site</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">minor_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
    <span class="n">minor_allele_map</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">minor_index</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;At site </span><span class="si">{</span><span class="n">site</span><span class="si">}</span><span class="s2">, the minor allele is </span><span class="si">{</span><span class="n">minor_allele_map</span><span class="p">[</span><span class="n">site</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>At site 0, the minor allele is G
At site 1, the minor allele is T
At site 2, the minor allele is T
At site 3, the minor allele is G
</pre></div>
</div>
</div>
</div>
<p>Given that information, we can recode the data to look like this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Warning: this code does not generalize to &gt; 2 alleles at a site,</span>
<span class="c1"># which is why the assert statement exists.</span>

<span class="n">recoded_matrix_minor_major</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span> <span class="n">site</span><span class="p">]))</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span> <span class="n">site</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">haplotypes</span><span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">site</span><span class="p">]</span> <span class="o">==</span> <span class="n">minor_allele_map</span><span class="p">[</span><span class="n">site</span><span class="p">]:</span>
            <span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

<span class="n">recoded_matrix_minor_major</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[0 0 0 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [0 0 0 1]
 [1 1 1 1]
 [1 1 1 0]]
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="the-site-or-allele-frequency-spectrum">
<h1>The site (or allele) frequency spectrum<a class="headerlink" href="#the-site-or-allele-frequency-spectrum" title="Permalink to this headline">¶</a></h1>
<div class="section" id="with-respect-to-the-derived-allele">
<h2>With respect to the derived allele<a class="headerlink" href="#with-respect-to-the-derived-allele" title="Permalink to this headline">¶</a></h2>
<p>To get the frequency spectrum, we simply tally the counts for derived states.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">derived_state_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="n">unfolded_fs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">derived_state_map</span><span class="p">:</span>
        <span class="n">derived_state_map</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">derived_state</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">derived_state_map</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">derived_state</span><span class="p">)</span>

<span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span><span class="n">site</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">state</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">state</span> <span class="ow">in</span> <span class="n">derived_state_map</span><span class="p">[</span><span class="n">site</span><span class="p">]:</span>
            <span class="n">unfolded_fs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">j</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="n">unfolded_fs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[8, 8, 8, 1]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="c1"># See https://stackoverflow.com/questions/30112420/histogram-for-discrete-values-with-matplotlib</span>
<span class="c1"># for hints :)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">unfolded_fs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">unfolded_fs</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1.5</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Derived allele count&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of sites&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">unfolded_fs</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))]);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/descriptive_statistics_33_0.png" src="../../_images/descriptive_statistics_33_0.png" />
</div>
</div>
</div>
<div class="section" id="with-respect-to-the-minor-allele">
<h2>With respect to the minor allele<a class="headerlink" href="#with-respect-to-the-minor-allele" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">folded_fs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">haplotypes</span><span class="p">[:,</span><span class="n">site</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">minor_count</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="n">where_minor_count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">minor_count</span><span class="p">)</span>
    <span class="n">folded_fs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">minor_count</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site index </span><span class="si">{</span><span class="n">site</span><span class="si">}</span><span class="s2">: minor allele = </span><span class="si">{</span><span class="n">counts</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">where_minor_count</span><span class="p">]</span><span class="si">}</span><span class="s2">, minor allele count = </span><span class="si">{</span><span class="n">minor_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site index 0: minor allele = [&#39;G&#39;], minor allele count = 2
Site index 1: minor allele = [&#39;T&#39;], minor allele count = 2
Site index 2: minor allele = [&#39;T&#39;], minor allele count = 2
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site index 3: minor allele = [&#39;G&#39;], minor allele count = 1
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="c1"># See https://stackoverflow.com/questions/30112420/histogram-for-discrete-values-with-matplotlib</span>
<span class="c1"># for hints :)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">folded_fs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">folded_fs</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1.5</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Minor allele count&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of sites&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">//</span> <span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">haplotypes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)]);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">folded_fs</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))]);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/descriptive_statistics_36_0.png" src="../../_images/descriptive_statistics_36_0.png" />
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="correlations-amongst-variable-sites-linkage-disequilibrium">
<h1>Correlations amongst variable sites (“linkage disequilibrium”)<a class="headerlink" href="#correlations-amongst-variable-sites-linkage-disequilibrium" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-first-of-perhaps-many-d-statistic-s">
<h2>The (first of perhaps many) <span class="math notranslate nohighlight">\(D\)</span> statistic(s).<a class="headerlink" href="#the-first-of-perhaps-many-d-statistic-s" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="pearson-s-squared-correlation-coefficient">
<h2>Pearson’s squared correlation coefficient<a class="headerlink" href="#pearson-s-squared-correlation-coefficient" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">site_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">site_j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">site_i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">major</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="p">[:,</span><span class="n">site_i</span><span class="p">],</span> <span class="n">recoded_matrix_minor_major</span><span class="p">[:,</span><span class="n">site_j</span><span class="p">])</span>
        <span class="n">ad</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">recoded_matrix_ancestral_derived</span><span class="p">[:,</span><span class="n">site_i</span><span class="p">],</span> <span class="n">recoded_matrix_ancestral_derived</span><span class="p">[:,</span><span class="n">site_j</span><span class="p">])</span>
        <span class="n">major_r2</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">major</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
        <span class="n">ad_r2</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">ad</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site pair (</span><span class="si">{</span><span class="n">site_i</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">site_j</span><span class="si">}</span><span class="s2">):&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">r^2 between major alleles = </span><span class="si">{</span><span class="n">major_r2</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">For derived alleles, r^2 = </span><span class="si">{</span><span class="n">ad_r2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site pair (0, 1):
	r^2 between major alleles = 1.0.
	For derived alleles, r^2 = 1.0
Site pair (0, 2):
	r^2 between major alleles = 1.0.
	For derived alleles, r^2 = 1.0
Site pair (0, 3):
	r^2 between major alleles = 0.0278.
	For derived alleles, r^2 = 0.0278
Site pair (1, 2):
	r^2 between major alleles = 1.0.
	For derived alleles, r^2 = 1.0
Site pair (1, 3):
	r^2 between major alleles = 0.0278.
	For derived alleles, r^2 = 0.0278
Site pair (2, 3):
	r^2 between major alleles = 0.0278.
	For derived alleles, r^2 = 0.0278
</pre></div>
</div>
</div>
</div>
<p>You will note that the result is the same for the two encodings of the data.
The reason is that the statistic is <em>squared</em>, which hides the fact that, for some sites, the correlation coefficient is equal in magnitude but opposite in sign for the two encodings:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">site_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="k">for</span> <span class="n">site_j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">site_i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">recoded_matrix_minor_major</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">major</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">recoded_matrix_minor_major</span><span class="p">[:,</span><span class="n">site_i</span><span class="p">],</span> <span class="n">recoded_matrix_minor_major</span><span class="p">[:,</span><span class="n">site_j</span><span class="p">])</span>
        <span class="n">ad</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">recoded_matrix_ancestral_derived</span><span class="p">[:,</span><span class="n">site_i</span><span class="p">],</span> <span class="n">recoded_matrix_ancestral_derived</span><span class="p">[:,</span><span class="n">site_j</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Site pair (</span><span class="si">{</span><span class="n">site_i</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">site_j</span><span class="si">}</span><span class="s2">):&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">r between major alleles = </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">major</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">For derived alleles, r = </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">ad</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site pair (0, 1):
	r between major alleles = 1.0.
	For derived alleles, r = 1.0
Site pair (0, 2):
	r between major alleles = 1.0.
	For derived alleles, r = 1.0
Site pair (0, 3):
	r between major alleles = -0.1667.
	For derived alleles, r = 0.1667
Site pair (1, 2):
	r between major alleles = 1.0.
	For derived alleles, r = 1.0
Site pair (1, 3):
	r between major alleles = -0.1667.
	For derived alleles, r = 0.1667
Site pair (2, 3):
	r between major alleles = -0.1667.
	For derived alleles, r = 0.1667
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./markdown/describing_variation_data"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="variation_data_intro.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Population genetic data</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="real_data_complications.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">The complexities of real data</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By Kevin R. Thornton<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>